Verified metrics

romeovs opened the pull request:
Closes https://github.com/metabase/metabase/issues/47803

### Description

Adds support for verified metrics to the browse metrics page in Enterprise Metabase.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Run Metabase EE and make sure you have a token set up with the `content_verification` feature
2. Make sure you have two metrics
3. Navigate to `/browse/metrics` by clicking `Metrics` in the sidenav
4. Navigate to one of the metrics
5. On the metric page click the "..." menu and click "Verify this metric"
6. Navigate back to `/browse/metrics` by clicking `Metrics` in the sidenav
7. A filter option should now be availble in the top right, play around with selecting and unselecting it to makes sure it works by filtering the metrics
8. Unverify the verified metric and make sure the filter on `/browse/metrics` is gone again

### Demo

https://github.com/user-attachments/assets/98f8bf4e-505f-4412-8213-a768ca62881a

romeovs committed:
Support skipToken in useFetchMetrics

romeovs committed:
WIP

romeovs committed:
Add user setting for verified metrics

romeovs committed:
Add verified metrics to the VERIFIED_CONTENT plugin

romeovs committed:
Use metric plugins for verified metrics

romeovs committed:
Move helpers down

romeovs committed:
Do not filter for verified metrics if there are none, regardless of user setting

romeovs committed:
Add label to filter button

romeovs committed:
Revalidate search when card gets (un-)verified

romeovs committed:
Add e2e test for verified metrics

romeovs committed:
Avoid checking for verified metrics on when content-verification is not enabled

romeovs committed:
Fix broken test

ranquild commented on enterprise/frontend/src/metabase-enterprise/moderation/actions.js:
```diff
@@ -13,6 +15,7 @@ export const verifyCard = createThunkAction(
       text,
     });
 
+    dispatch(Api.util.invalidateTags([listTag("card")]));
```
We need to change `verifyItem` and possibly `removeReview` to use RTK query mutations (via redux actions, like we do in entities). And these mutations should invalidate `card`s, as usual.

ranquild reviewed (commented):
I think this is currently implemented in a very roundabout way and it be done much simpler. Something like this:
```
type MetricFilters = {
  isVerified: boolean;
}

type MetricFilterControlsProps = {
  filters: MetricFilters;
  onChange: (newFilters: MetricFilters) => void;
};

// OSS defaults
const PLUGIN_CONTENT_VERIFICATION = {
  ...
  getDefaultMetricFilters: (_state: State) => ({ isVerified: false });
  MetricFilterControls: (_props: MetricFilterControlsProps): JSX.Element | null => null;
};

// EE
PLUGIN_CONTENT_VERIFICATION.getDefaultMetricFilters = (state) => ({
  isVerified: getSetting(state, "....")
})
PLUGIN_CONTENT_VERIFICATION.MetricFilterControls = MetricFilterControls;
```

so you define the filter component in EE and use it on OSS. You shouldn't check for any token features in OSS code and always rely on the plugin doing the right thing.

romeovs committed:
Merge branch 'master' into 47803-verified-metrics

romeovs committed:
Simplify content verification plugin structure for metrics

romeovs committed:
Add content moderation types and API helpers

romeovs committed:
Update content moderation plugin to make use of the moderation api in metabase/api

romeovs committed:
Fix lint

romeovs committed:
Fix missing list tag

romeovs commented:
@ranquild I've simplified the plugin a little bit and added an content moderation to `metabase/api`.

ranquild reviewed (commented):
@romeovs everything is great, but (!) the user setting is not saved now, ever. I think you could update it in `MetricFilterControls` when the value changes. Please fix this + add a e2e test for this case.

romeovs committed:
Store verified metric filter preference

romeovs committed:
Move all helpers to the bottom

romeovs committed:
Ensure the user setting gets updated when chaning the filter

rafpaf commented on enterprise/frontend/src/embedding-sdk/CHANGELOG.md:
This question might be off-base, but I am curious, should these changes be included in this PR? Seeing this made me wonder if an IDE formatter, like prettier, adjusted the formatting of the bullet points.

romeovs replied:
This is due to the branch being out of date

rafpaf replied:
On master I see `*` asterisks in this file. If this PR gets merged, I think the file will change, unless I'm mistaken.

rafpaf commented on enterprise/frontend/src/embedding-sdk/README.md:
My guess is that this might be a whitespace change added by an IDE formatter?

romeovs replied:
This is due to the branch being out of date.

rafpaf commented on enterprise/frontend/src/metabase-enterprise/content_verification/metrics.tsx:
```diff
@@ -0,0 +1,83 @@
+
+  const handleVerifiedFilterChange = useCallback(
+    function (evt: ChangeEvent<HTMLInputElement>) {
+      setMetricFilters({ ...metricFilters, verified: evt.target.checked });
```
In `enterprise/frontend/src/metabase-enterprise/content_verification/ModelFilterControls.tsx`, the function `setVerifiedFilterStatus`, which I think is analogous to `setMetricFilters`, is debounced. This avoids a problem where clicking the filter a few times in a row makes the page behave oddly. Might it be worth debouncing `setMetricFilters` for the same reason?

romeovs replied:
I'm not seeing any odd behavior when clicking the filter a few times (very fast even)! Could it be that the issue is not present for Metrics?

ranquild replied:
yeah the metric implementation is slightly different and doesn't have this issue

rafpaf replied:
Excellent!

rafpaf commented on enterprise/frontend/src/metabase-enterprise/content_verification/metrics.tsx:
```diff
@@ -0,0 +1,83 @@
+  };
+}
+
+export const MetricFilterControls = ({
```
You could add a comment here noting that this component is very similar to `ModelFilterControls`. I don't think it's necessary just yet to unify the two components, but it might eventually become helpful to do that – for example, if a similar filtering mechanism is needed for a third kind of entity besides models and metrics. A comment like that might help a future dev reason about this unification.

rafpaf commented on frontend/src/metabase/static-viz/components/ComboChart/stories-data/bar-breakout-with-line-series-stacked-right-axis-only.json:
I'm not sure what the context of these changes is

romeovs replied:
I think the branch was too out of date. I've merged in `master` so it should be clearer now.

rafpaf replied:
I still see lots of unrelated changes here. Maybe `git checkout master -- $file` for each weird file?

rafpaf reviewed (changes requested):
Looks good. I've made some small suggestions.

Some whitespace changes seem to have snuck into this PR – I'm guessing due to an overeager IDE formatter.

romeovs committed:
Merge branch 'master' into 47803-verified-metrics

romeovs commented:
> @ranquild  everything is great, but (!) the user setting is not saved now, ever. I think you could update it in `MetricFilterControls` when the value changes. Please fix this + add a e2e test for this case.

I've added e2e tests for this and fixed the bug!

romeovs committed:
Add comment about component similarity

ranquild commented:
@rafpaf could you take another look?

romeovs committed:
Merge branch 'master' into 47803-verified-metrics

romeovs committed:
Undo changes to embedding-sdk
